### PR TITLE
Optimize CI tests

### PR DESCRIPTION
**Pull request by husio**
Improvements to CI setup to shorten the execution time. Current build time is [18 min 51 sec](https://travis-ci.com/iov-one/weave/builds/105310606).

- Cache `GOPATH` between builds
Before using cache (empty cache, first run): Total time 19 min 4 sec 
First time after using cache (GOPATH restored):  Total time 16 min 2 sec :chart_with_upwards_trend: 

- Run coverage and binary/container builds only for the main Go version (1.11 right now).
- Build the container only if it will be published.
- Do not remove `vendor/`  directory

[Total time 12 min 51 sec](https://travis-ci.com/iov-one/weave/builds/105429273)   :tada:  - 6min improvement.

Addint one more Go version (1.12) will not make this process any longer.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,19 +12,15 @@ GOPATH ?= $$HOME/go
 
 PROTOC_FLAGS := -I=. -I=./vendor -I=$(GOPATH)/src
 
-all: deps build test
+all: deps test
 
 dist:
-	cd cmd/bnsd ; make dist ; cd -
-	cd cmd/bcpd ; make dist ; cd -
+	cd cmd/bnsd && $(MAKE) dist
```
Comment (alpe) [inline Makefile]: 👍

**Comment by ethanfrey** [comment]
33% speedup.
Nice work.

File: .travis.yml [hunk 2]
```diff
@@ -16,51 +16,65 @@ os:
...
-  docker push "iov1/bnsd:$TRAVIS_TAG";
-  docker logout;
-  fi
+- make lint;
+- make test;
+- if [[ "$TRAVIS_GO_VERSION" == "$MAIN_GO_VERSION" ]]; then
+    make cover;
+    ./coverage/upload.sh;
+  fi;
+- if [[ "$TRAVIS_GO_VERSION" == "$MAIN_GO_VERSION" && "$TRAVIS_OS_NAME" == "linux" ]]; then
```
Comment (ethanfrey) [inline .travis.yml]: Nice cleanup here, but can you add it back to `after_success`.
Point is not to run it if tests fail

File: Makefile (hunk 1 above)
Comment (ethanfrey) [reply Makefile]: DIfference between $(MAKE) and make?

Comment (husio) [reply Makefile]: I guess it can be overwritten if you use non standard makefile :man_shrugging:

File: .travis.yml (hunk 2 above)
Comment (husio) [reply .travis.yml]: If any step fails then further steps are not called.

Here is an example:
- [change that is adding `exit 2`](https://github.com/iov-one/weave/pull/433/commits/e35b43a18121d117d511d57148d3c4f626ad6832) to the build process
- [CI build](https://travis-ci.com/iov-one/weave/jobs/187386755#L550) that does not run anything after `exit 2`

**Pull request merged**